src/backend/distributed/commands/function.c

```diff
@@ -91,6 +187,309 @@ CreateFunctionDDLCommandsIdempotent(const ObjectAddress *functionAddress)
 }
 
 
+/*
+ * GetDistributionArgIndex calculates the distribution argument with the given
+ * parameters. The function errors out if no valid argument is found.
+ */
+static int
+GetDistributionArgIndex(Oid functionOid, char *distributionArgumentName,
+						Oid *distributionArgumentOid)
+{
+	int distributionArgumentIndex = -1;
+
+	int numberOfArgs = 0;
+	int argIndex = 0;
+	Oid *argTypes = NULL;
+	char **argNames = NULL;
+	char *argModes = NULL;
+
+	HeapTuple proctup = NULL;
+
+	*distributionArgumentOid = InvalidOid;
+
+	proctup = SearchSysCache1(PROCOID, ObjectIdGetDatum(functionOid));
+	if (!HeapTupleIsValid(proctup))
+	{
+		elog(ERROR, "cache lookup failed for function %u", functionOid);
+	}
+
+	numberOfArgs = get_func_arg_info(proctup, &argTypes, &argNames, &argModes);
+
+	if (argumentStartsWith(distributionArgumentName, "$"))
+	{
+		/* skip the first character, we're safe because text_to_cstring pallocs */
+		distributionArgumentName++;
+
+		/* throws error if the input is not an integer */
+		distributionArgumentIndex = pg_atoi(distributionArgumentName, 4, 0);
+
+		if (distributionArgumentIndex < 1 || distributionArgumentIndex > numberOfArgs)
+		{
+			char *functionName = get_func_name(functionOid);
+
+			ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							errmsg("cannot distribute the function \"%s\" since "
+								   "the distribution argument is not valid",
+								   functionName),
+							errhint("Either provide a valid function argument name "
+									"or a valid \"$paramIndex\" to "
+									"create_distributed_function()")));
+		}
+
+		/*
+		 * Internal representation for the distributionArgumentIndex
+		 * starts from 0 whereas user facing API starts from 1.
+		 */
+		distributionArgumentIndex -= 1;
+	}
+
+	/*
+	 * If the user provided "$paramIndex", we're unnecessarily looping, but we prefer
+	 * this for the simplicity of the code.
+	 */
```

Comment:
When I read this I didn't understand why we needed to go into the loop at all. Maybe add that we still need to find the OID of the argument.

btw, wouldn't something like:
```c
*distributionArgumentOid = argTypes[distributionArgumentIndex];

/*
 * Internal representation for the distributionArgumentIndex
 * starts from 0 whereas user facing API starts from 1.
 */
return distributionArgumentIndex  - 1;
```

do the trick?

Reply:
> do the trick?

I guess no, so there are two cases:  (a) the user types in like function parameter: `$1` (b) gives the name of the parameter : `val1`

For (a) we don't need to loop over array, we can just use `argTypes[distributionArgumentIndex];`. However, for (b) we need to iterate over the arguments  and check whether the name exists in the parameters of the function. If we find, we can use `argTypes[distributionArgumentIndex]`, otherwise we error out.

So, the comment actually aims to  explain that. I expanded the comment a bit more to clarify.

Reply:
I meant in the `if (argumentStartsWith(distributionArgumentName, "$"))` block above .

Reply:
I see, yes, that seem to simplify the code. Let me do that =, thanks